source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc

```diff
@@ -322,6 +322,12 @@ Http::FilterHeadersStatus JsonTranscoderFilter::encodeHeaders(Http::HeaderMap& h
 
   response_headers_ = &headers;
 
+  // remove Trailer headers if the client connection was http/1
```

Comment:
when end_stream is false and the grpc call is not server streaming, the filter StopIteration until it receives trailer, so that's why this should be moved to `encodeTrailers`. do not remove the header when the call IS streaming (which is chunked encoding).

Reply:
`encodeTrailers` appears to _only_ be called when `end_stream` is true though:

```c++
  if (end_stream) {
    // In gRPC wire protocol, headers frame with end_stream is a trailers-only response.
    // The return value from encodeTrailers is ignored since it is always continue.
    encodeTrailers(headers);
    return Http::FilterHeadersStatus::Continue;
  }
```

I could add a check in `encodeHeaders` to ensure we are not removing the Trailer header if the request is streaming. Would that work?

Reply:
No, that is not the case. In a regular (non trailers-only) gRPC response, Envoy will call `encodeHeaders`, `encodeData`, `encodeTrailers` sequentially when frames are received from upstream. The transcoder filter do return `StopIteration` in first two so Envoy will buffer them, until it reaches `encodeTrailers` for non-streaming cases. This is why the transcoder could set `content-length` [here](https://github.com/envoyproxy/envoy/blob/master/source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc#L401) for header, even in `encodeTrailers`, so what you want to do is remove trailers at the same place.